core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/HardenedMarecoAllowance.java

```diff
@@ -104,46 +115,60 @@ private Envelope computeMarecoCore(Envelope coreBase, double v1) {
         return res;
     }
 
-    private EnvelopePart computeSlowdown(Envelope base, double v1) {
-        var startSpeed = base.getBeginSpeed();
-        if (startSpeed <= v1)
+    private EnvelopePart computeSlowdown(Envelope envelopeSection, double v1, double targetBeginSpeed) {
+        double beginSpeed;
+        // if no begin speed has been imposed by the previous range
+        if (targetBeginSpeed == 0) {
+            // if the section is not at the beginning of the allowance region, there will be no slowdown phase
+            if (envelopeSection.getBeginPos() > beginPos)
+                return null;
+            beginSpeed = envelopeSection.getBeginSpeed();
+        } else
+            beginSpeed = targetBeginSpeed;
+        if (beginSpeed <= v1)
             return null;
-
         var partBuilder = new EnvelopePartBuilder();
         var constrainedBuilder = new ConstrainedEnvelopePartBuilder(
                 partBuilder,
                 new SpeedFloor(v1)
         );
-        EnvelopeDeceleration.decelerate(context, base.getBeginPos(), startSpeed, constrainedBuilder, 1);
+        EnvelopeDeceleration.decelerate(context, envelopeSection.getBeginPos(), beginSpeed, constrainedBuilder, 1);
         return partBuilder.build();
     }
 
-    private EnvelopePart computeSpeedup(Envelope base, double v1) {
-        var endSpeed = base.getEndSpeed();
+    private EnvelopePart computeSpeedup(Envelope envelopeSection, double v1, double targetEndSpeed) {
+        double endSpeed;
+        // if no begin speed has been imposed by the next range
+        if (targetEndSpeed == 0) {
```

Comment:
0 means unspecified? I guess it makes sense if we consider that it can't be an imposed speed, but I'd prefer NaN if that works.